torch/utils/_pytree.py

```diff
@@ -847,30 +873,59 @@ def unflatten(self, leaves: Iterable[Any]) -> PyTree:
         start = 0
         end = 0
         child_pytrees = []
-        for child_spec in self.children_specs:
+        for child_spec in self._children:
             end += child_spec.num_leaves
             child_pytrees.append(child_spec.unflatten(leaves[start:end]))
             start = end
 
-        return unflatten_fn(child_pytrees, self.context)
+        return unflatten_fn(child_pytrees, self._context)
 
 
+@deprecated(
+    "`isinstance(treespec, LeafSpec)` is deprecated, "
+    "use `isinstance(treespec, TreeSpec)` and `treespec.is_leaf()` instead.",
+    category=FutureWarning,
+)
 class LeafSpec(TreeSpec):
     def __init__(self) -> None:
         super().__init__(None, None, [])
 
-    def __post_init__(self) -> None:
-        object.__setattr__(self, "num_nodes", 1)
-        object.__setattr__(self, "num_leaves", 1)
-        object.__setattr__(self, "num_children", 0)
-
     def __repr__(self, indent: int = 0) -> str:
         return "*"
 
 
 # All leaves are equivalent, so represent with a single object to save on
 # object construction time
-_LEAF_SPEC = LeafSpec()
+with warnings.catch_warnings():
+    warnings.filterwarnings(
+        "ignore", category=FutureWarning, module=__name__, append=False
+    )
+    _LEAF_SPEC = LeafSpec()
+
+
+def treespec_leaf() -> LeafSpec:
+    """Make a treespec representing a leaf node."""
+    return _LEAF_SPEC
```

Comment:
Is this supposed to be public API? What's the rationale behind this change?

Reply:
I think it could be public to provide a standard way to create `TreeSpec` instances without exposing the implementation details. `treespec_{leaf,tuple,dict}` are only used in `torch.export`, it could be private for now. But I think explicit public APIs are the better solution.